xdsl/dialects/memref.py

```diff
@@ -587,6 +587,33 @@ def get(
         return Cast.build(operands=[source], result_types=[type])
 
 
+@irdl_op_definition
+class MemorySpaceCast(IRDLOperation):
+    name = "memref.memory_space_cast"
+
+    source: Operand = operand_def(MemRefType[Attribute] | UnrankedMemrefType[Attribute])
+    dest: OpResult = result_def(MemRefType[Attribute] | UnrankedMemrefType[Attribute])
+
+    def __init__(
+        self,
+        source: SSAValue | Operation,
+        dest: MemRefType[Attribute] | UnrankedMemrefType[Attribute],
+    ):
+        super().__init__(operands=[source], result_types=[dest])
```

Comment:
If I understand correctly, the only thing that could change in the result type is the memory space, right?
If so, I would make the constructor just take the new memory space and construct the corresponding MemrefType from the input one, for user-friendliness

(Just an idea, not a requirement for merging at all :slightly_smiling_face: )

Reply:
Sorry to give conflicting advice, but I have a feeling that there was a benefit of the previous version, in that it seems more flexible. The new initialiser might not be usable in the context of a rewrite, where the type of the ssa value being passed might be invalid, and fixed later on in the compiler pass. I would recommend reverting the latest change, and adding the new function either as a helper, or as an overload to the init, like in `arith.constant`

Reply:
this is great, init + constructor seems like the best way to go for now